wren-ui/src/apollo/server/managers/dataSourceSchemaDetector.ts

```diff
@@ -0,0 +1,288 @@
+import { camelCase, differenceWith, isEmpty, isEqual } from 'lodash';
+import { CompactTable } from '@server/connectors/connector';
+import { DataSourceStrategyFactory } from '@server/factories/onboardingFactory';
+import { IContext } from '@server/types';
+import { getLogger } from 'log4js';
+
+const logger = getLogger('DataSourceSchemaDetector');
+logger.level = 'debug';
+
+export type DataSourceSchema = {
+  name: string;
+  columns: {
+    name: string;
+    type: string;
+  }[];
+};
+
+export type DataSourceSchemaChange = {
+  [SchemaChangeType.DELETED_TABLES]?: DataSourceSchema[];
+  [SchemaChangeType.DELETED_COLUMNS]?: DataSourceSchema[];
+  [SchemaChangeType.MODIFIED_COLUMNS]?: DataSourceSchema[];
+};
+
+export type DataSourceSchemaResolve = {
+  [SchemaChangeType.DELETED_TABLES]?: boolean;
+  [SchemaChangeType.DELETED_COLUMNS]?: boolean;
+  [SchemaChangeType.MODIFIED_COLUMNS]?: boolean;
+};
+
+export enum SchemaChangeType {
+  // the tables has been deleted
+  DELETED_TABLES = 'deletedTables',
+  // the columns has been deleted
+  DELETED_COLUMNS = 'deletedColumns',
+  // the columns type has been changed
+  MODIFIED_COLUMNS = 'modifiedColumns',
+}
+
+export interface IDataSourceSchemaDetector {
+  detectSchemaChange(): Promise<void>;
+  resolveSchemaChange(type: string): Promise<void>;
+}
+
+export default class DataSourceSchemaDetector
+  implements IDataSourceSchemaDetector
+{
+  public ctx: IContext;
+  public projectId: number;
+
+  constructor({ ctx, projectId }: { ctx: IContext; projectId: number }) {
+    this.ctx = ctx;
+    this.projectId = projectId;
+  }
+
+  public async detectSchemaChange() {
+    const diffSchema = await this.getDiffSchema();
+    if (diffSchema) {
+      this.addSchemaChange(diffSchema);
+    }
+  }
+
+  public async resolveSchemaChange(type: string) {
+    const schemaChangeType = camelCase(type) as SchemaChangeType;
+    const supportedTypes = [
+      SchemaChangeType.DELETED_TABLES,
+      SchemaChangeType.DELETED_COLUMNS,
+    ];
+    if (supportedTypes.includes(schemaChangeType)) {
+      const lastSchemaChange =
+        await this.ctx.schemaChangeRepository.findLastSchemaChange(
+          this.projectId,
+        );
+      const changes = lastSchemaChange?.change[schemaChangeType];
+      const isResolved = lastSchemaChange?.resolve[schemaChangeType];
+
+      if (isResolved !== false) {
```

Comment:
```suggestion
      if (isResolved === true) {
```
I think it is easier to read =D

Reply:
I'll also remove the `[key]: null` in resolve since it does not satisfy this condition.